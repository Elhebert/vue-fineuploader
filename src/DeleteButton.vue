<template>
  <renderless-delete-button
    :id="id"
    :uploader="uploader">
    <button
      v-if="!onlyRenderIfDeletable || deletable || deleting"
      slot-scope="{ deleting, deletable, buttonEvents }"
      :disabled="deleting || !deletable"
      type="button"
      v-on="buttonEvents"
    >
      Delete
    </button>
  </renderless-delete-button>
</template>

<script>
import RenderlessDeleteButton from './renderless/DeleteButton'

export default {
  components: {
    RenderlessDeleteButton,
  },

  props: {
    id: {
      type: Number,
      required: true,
    },
    onlyRenderIfDeletable: {
      type: Boolean,
      default: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },
}
</script>
