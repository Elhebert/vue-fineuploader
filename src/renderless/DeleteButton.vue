<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  data: () => ({
    deletable: false,
    deleting: false,
    unmounted: false,
  }),

  created() {
    this.uploader.qq.status = this.uploader.qq.status
    const initialStatus = this.uploader.methods.getUploads({
      id: this.id,
    }).status

    this.deletable = this.isDeletableStatus(initialStatus)
    this.deleting = initialStatus === this.uploader.qq.status.DELETING
  },

  mounted() {
    this.uploader.on('statusChange', this.onStatusChange)
  },

  beforeDestroy() {
    this.unmounted = true
    this.unregisterStatusChangeHandler()
  },

  methods: {
    onStatusChange(id, oldStatus, newStatus) {
      if (!this.isCurrentFile(id) || this.unmounted) {
        return
      }

      if (
        !this.isDeletableStatus(newStatus) &&
        newStatus !== this.uploader.qq.status.DELETING
      ) {
        this.deletable = false
        this.deleting = false
        this.unregisterStatusChangeHandler()

        return
      }

      if (this.isDeletableStatus(newStatus)) {
        this.deletable = true
        this.deleting = false

        return
      }

      if (newStatus === this.uploader.qq.status.DELETING) {
        this.deleting = true
      }
    },

    isCurrentFile(id) {
      return this.id === id
    },

    onClick(event) {
      event.preventDefault()

      if (this.deletable && !this.deleting) {
        this.uploader.methods.deleteFile(this.id)
      }
    },

    unregisterStatusChangeHandler() {
      this.uploader.off('statusChange', this.onStatusChange)
    },

    isDeletableStatus(statuses) {
      return (
        [
          this.uploader.qq.status.DELETE_FAILED,
          this.uploader.qq.status.UPLOAD_SUCCESSFUL,
        ].indexOf(statuses) >= 0
      )
    },
  },

  render() {
    return this.$scopedSlots.default({
      buttonEvents: {
        click: event => {
          this.onClick(event)
        },
      },
      deletable: this.deletable,
      deleting: this.deleting,
    })
  },
}
</script>
