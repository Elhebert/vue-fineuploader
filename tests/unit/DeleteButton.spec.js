import { mount } from '@vue/test-utils'
import DeleteButton from '@/DeleteButton.vue'
import FineUploader from 'fine-uploader-wrappers'
import { status } from 'fine-uploader/lib/core/all'

describe('DeleteButton.vue', () => {
  let wrapper
  let renderless
  let uploader

  beforeEach(() => {
    uploader = new FineUploader({ options: {} })

    wrapper = mount(DeleteButton, {
      propsData: { id: 0, uploader },
    })

    renderless = wrapper.vm.$children[0]
  })

  it('renders the button for a successfully uploaded file', () => {
    renderless.onStatusChange(0, null, status.UPLOAD_SUCCESSFUL)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toBeDefined()
    expect(wrapper.attributes().disabled).toBeUndefined()
  })

  it('deletes the file if clicked', () => {
    uploader.methods.deleteFile = jest.fn()

    renderless.onStatusChange(0, null, status.UPLOAD_SUCCESSFUL)

    wrapper.trigger('click')
    expect(uploader.methods.deleteFile).toHaveBeenCalled()
  })

  it('removes the button if the file can no longer be deleted', () => {
    renderless.onStatusChange(0, null, status.DELETED)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.isEmpty()).toBe(true)
  })

  it('disabled the button while the delete is in progress', () => {
    renderless.onStatusChange(0, null, status.DELETING)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().disabled).toBeDefined()
  })

  it('disables the button if requested when the file can no longer be deleted', () => {
    wrapper.setProps({ onlyRenderIfDeletable: false })
    renderless.onStatusChange(0, null, status.DELETED)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toBeDefined()
    expect(wrapper.attributes().disabled).toBeDefined()
  })
})
