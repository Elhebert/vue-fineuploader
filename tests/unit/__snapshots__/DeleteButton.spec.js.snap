// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteButton.vue disabled the button while the delete is in progress 1`] = `
<button disabled="disabled" type="button">
  Delete
</button>
`;

exports[`DeleteButton.vue disables the button if requested when the file can no longer be deleted 1`] = `
<button disabled="disabled" type="button">
  Delete
</button>
`;

exports[`DeleteButton.vue removes the button if the file can no longer be deleted 1`] = ``;

exports[`DeleteButton.vue renders the button for a successfully uploaded file 1`] = `
<button type="button">
  Delete
</button>
`;
